bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/impl/ProductGraphqlServicImpl.java

```diff
@@ -0,0 +1,151 @@
+/*
+ *   Copyright 2019 Adobe Systems Incorporated
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.adobe.cq.commerce.core.components.internal.impl;
+
+import com.adobe.cq.commerce.core.components.internal.models.v1.product.AssetImpl;
+import com.adobe.cq.commerce.core.components.models.product.Asset;
+import com.adobe.cq.commerce.core.components.service.ProductGraphqlService;
+import com.adobe.cq.commerce.graphql.client.GraphqlClient;
+import com.adobe.cq.commerce.graphql.client.GraphqlRequest;
+import com.adobe.cq.commerce.graphql.client.GraphqlResponse;
+import com.adobe.cq.commerce.magento.graphql.*;
+import com.adobe.cq.commerce.magento.graphql.gson.Error;
+import com.adobe.cq.commerce.magento.graphql.gson.QueryDeserializer;
+import org.osgi.service.component.annotations.Component;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ *
+ *GraphQlService extracted from ProductModel
+ * It could be one service to cater to all component , carrying all the service methods
+ * ToDO: Refactoring
+ *
+ */
+
+@Component(service = ProductGraphqlService.class)
```

Comment:
We haven't used a service for requests until now, exactly because each component is responsible to request exactly what it needs. For example, the product teaser might need less product attributes that the product detail, so the idea is to get *exactly* what you need. I'd hence rather move that to the Sling model, like we did for all other components. I think you should strip down the GraphQL request and *only* request the fields needed by this component. For example, don't query `id` or `categories` if you don't use them.

Note that you could still reuse some methods implemented elsewhere, like `ProductImpl.filterAndSortAssets()` which you can change to `public`, it's anyway not OSGi exported so it's not a problem.

Reply:
I agree with @cjelger the introduction of the services changes the way how components worked before as they shift the responsibility of data fetching to the service now. Let's discuss that approach in general.

Reply:
Ok, I got the idea here. I was hoping to suggest on putting all graphql related queries in one generic service later  to somehow reduce code duplicity. but yeah requesting the only required fields is better use of graphql. Moving this to Model now, with more restrictive query.

Reply:
The "generic service" will only be a big container as most components might request different attributes and use different queries. But yes some generic parts like the subpart to query a price can certainly be reused.